azure/functions/extension/app_extension_base.py

```diff
@@ -0,0 +1,122 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import typing
+from logging import Logger
+from .extension_meta import ExtensionMeta
+from .extension_scope import ExtensionScope
+from .._abc import Context
+
+
+class AppExtensionBase(metaclass=ExtensionMeta):
+    """An abstract class defines the global life-cycle hooks to be implemented
+    by customer's extension, will be applied to all functions.
+
+    An AppExtension should be treated as a static class. Must not contain
+    __init__ method since it is not instantiable.
+
+    Please place your initialization code in init() classmethod, consider
+    accepting extension settings in configure() classmethod from customers.
+    """
+
+    _scope = ExtensionScope.APPLICATION
+
+    @classmethod
+    def init(cls):
+        """The function will be executed when the extension is loaded.
+        Happens when Azure Functions customers import the extension module.
+        """
+        pass
+
+    @classmethod
+    def configure(cls, *args, **kwargs):
+        """This function is intended to be called by Azure Functions
+        customers. This is a contract between extension developers and
+        azure functions customers. If multiple .configure() are called,
+        the extension system cannot guarentee the calling order.
+        """
+        pass
+
+    # DO NOT decorate this with @abc.abstractstatismethod
+    # since implementation by subclass is not mandatory
+    @classmethod
+    def post_function_load_app_level(cls,
```

Comment:
Yeah, we don't support **pre_function_load_app_level**, because the extension needs to be imported by customer's code (e.g. after the Python worker loaded their function.).

The **post_function_load_app_level** is similar to the **init** interface, but it provides function name and directory.

Basically, when an extension do initialization which require specific loading on trigger directory (e.g. reading customer's function.json), the extension will need to implement **post_function_load_app_level** instead.